modules/lib/src/main/resources/assets/js/app/widget/WidgetsToolbar.ts

```diff
@@ -0,0 +1,160 @@
+import {DivEl} from '@enonic/lib-admin-ui/dom/DivEl';
+import {Button} from '@enonic/lib-admin-ui/ui/button/Button';
+import {AppContext} from '../AppContext';
+import {Widget} from '@enonic/lib-admin-ui/content/Widget';
+import {ImgEl} from '@enonic/lib-admin-ui/dom/ImgEl';
+import {WidgetButton} from './WidgetButton';
+
+export class WidgetsToolbar
+    extends DivEl {
+
+    private buttons: WidgetButton[] = [];
+
+    private itemSelectedListeners: { (appOrWidgetId: string): void } [] = [];
+
+    private static MAIN_APP_ENDING: string = 'studio:main';
+
+    private static ARCHIVE_APP_ENDING: string = 'plus:archive';
+
+    private static SETTINGS_APP_ENDING: string = 'studio:settings';
+
+    constructor() {
+        super('actions-block');
+    }
+
+    addWidget(widget: Widget, buttonClass?: string): void {
+        this.createWidgetButton(widget, buttonClass);
+    }
+
+    private createWidgetButton(widget: Widget, buttonClass?: string) {
+        const sidebarButton: WidgetButton = new WidgetButton(widget);
+        sidebarButton.setLabel(widget.getDisplayName());
+        sidebarButton.setTitle(widget.getDisplayName());
+
+        if (buttonClass) {
+            sidebarButton.addClass(buttonClass);
+        }
+
+        const imgEl: ImgEl = new ImgEl(widget.getIconUrl());
+        sidebarButton.appendChild(imgEl);
+
+        this.listenButtonClicked(sidebarButton);
+
+        const pos: number = this.getButtonPos(sidebarButton);
+
+        if (pos < 0) {
+            this.appendChild(sidebarButton);
+        } else {
+            this.insertChild(sidebarButton, pos);
+        }
+
+        this.buttons.push(sidebarButton);
+    }
+
+    toggleActiveButton() {
+        this.buttons.forEach((b: WidgetButton) => {
+            b.toggleSelected(b.getWidgetId() === AppContext.get().getCurrentAppOrWidgetId());
+        });
+    }
+
+    private getButtonPos(sidebarButton: WidgetButton): number {
+        if (this.isMainAppButton(sidebarButton)) {
+            return 0;
+        }
+
+        if (this.isArchiveAppButton(sidebarButton)) {
+            return 1;
+        }
+
+        if (this.isSettingsApp(sidebarButton)) {
+            return -1;
+        }
+
+        return this.getNonDefaultAppPos(sidebarButton);
+    }
+
+    private getNonDefaultAppPos(newButton: WidgetButton): number {
+        let pos: number = -1;
+
+        this.getChildren().some((existingButton: WidgetButton, index: number) => {
+            if (this.isMainAppButton(existingButton) || this.isArchiveAppButton(existingButton)) {
+                return false;
+            }
+
+            if (this.isSettingsApp(existingButton)) { // inserting before Settings button
+                pos = index;
+                return true;
+            }
+
+            // finding
+            if (newButton.getWidgetDisplayName().toLowerCase().localeCompare(existingButton.getWidgetDisplayName().toLowerCase()) < 0) {
```

Comment:
`newButton.getWidgetDisplayName().toLowerCase()` should be called once before `.some()`. Actually, we don't even need entire `newButton` in this function, we only need its display name.